File: problem11/perl/matrix_product.pl
use strict;
use warnings;

# Returns ($max_product, $coords)
sub greatest_product {
    my ($matrix, $adj) = @_;
    my $rows = scalar(@$matrix);
    my $cols = scalar(@{$matrix->[0]});
    my $max = 0;
    my $best_coords = [];
    for my $i (0..$rows-1) {
        for my $j (0..$cols-1) {
            # right
            if ($j + $adj <= $cols) {
                my $prod = 1;
                my @coords;
                for my $k (0..$adj-1) {
                    $prod *= $matrix->[$i][$j+$k];
                    push @coords, [$i, $j+$k];
                }
                if ($prod > $max) {
                    $max = $prod;
                    $best_coords = [@coords];
                }
            }
            # down
            if ($i + $adj <= $rows) {
                my $prod = 1;
                my @coords;
                for my $k (0..$adj-1) {
                    $prod *= $matrix->[$i+$k][$j];
                    push @coords, [$i+$k, $j];
                }
                if ($prod > $max) {
                    $max = $prod;
                    $best_coords = [@coords];
                }
            }
            # diag down-right
            if ($i + $adj <= $rows && $j + $adj <= $cols) {
                my $prod = 1;
                my @coords;
                for my $k (0..$adj-1) {
                    $prod *= $matrix->[$i+$k][$j+$k];
                    push @coords, [$i+$k, $j+$k];
                }
                if ($prod > $max) {
                    $max = $prod;
                    $best_coords = [@coords];
                }
            }
            # diag down-left
            if ($i + $adj <= $rows && $j - $adj + 1 >= 0) {
                my $prod = 1;
                my @coords;
                for my $k (0..$adj-1) {
                    $prod *= $matrix->[$i+$k][$j-$k];
                    push @coords, [$i+$k, $j-$k];
                }
                if ($prod > $max) {
                    $max = $prod;
                    $best_coords = [@coords];
                }
            }
        }
    }
    return ($max, $best_coords);
}

1;

File: problem11/swift/MatrixProduct.swift
public func greatestProduct(matrix: [[Int]], adj: Int) -> (Int, [(Int, Int)]) {
    let rows = matrix.count
    let cols = matrix.first?.count ?? 0
    var maxProduct = 0
    var bestCoords: [(Int, Int)] = []
    for i in 0..<rows {
        for j in 0..<cols {
            // right
            if j + adj <= cols {
                var prod = 1
                var coords: [(Int, Int)] = []
                for k in 0..<adj {
                    prod *= matrix[i][j + k]
                    coords.append((i, j + k))
                }
                if prod > maxProduct {
                    maxProduct = prod
                    bestCoords = coords
                }
            }
            // down
            if i + adj <= rows {
                var prod = 1
                var coords: [(Int, Int)] = []
                for k in 0..<adj {
                    prod *= matrix[i + k][j]
                    coords.append((i + k, j))
                }
                if prod > maxProduct {
                    maxProduct = prod
                    bestCoords = coords
                }
            }
            // diag down-right
            if i + adj <= rows && j + adj <= cols {
                var prod = 1
                var coords: [(Int, Int)] = []
                for k in 0..<adj {
                    prod *= matrix[i + k][j + k]
                    coords.append((i + k, j + k))
                }
                if prod > maxProduct {
                    maxProduct = prod
                    bestCoords = coords
                }
            }
            // diag down-left
            if i + adj <= rows && j >= adj - 1 {
                var prod = 1
                var coords: [(Int, Int)] = []
                for k in 0..<adj {
                    prod *= matrix[i + k][j - k]
                    coords.append((i + k, j - k))
                }
                if prod > maxProduct {
                    maxProduct = prod
                    bestCoords = coords
                }
            }
        }
    }
    return (maxProduct, bestCoords)
}
